lib/download.go

```diff
@@ -1,47 +1,137 @@
 package lib
 
 import (
+	"errors"
 	"fmt"
 	"io"
+	"log"
 	"net/http"
 	"os"
 	"path/filepath"
 	"strings"
 )
 
-// DownloadFromURL : Downloads the binary from the source url
-func DownloadFromURL(installLocation string, url string) (string, error) {
+var (
+	PubKeyId     = "72D7468F"
+	PubKeyPrefix = "hashicorp_"
+	PubKeyUri    = "https://www.hashicorp.com/.well-known/pgp-key.txt"
+)
+
+const (
+	pubKeySuffix = ".asc"
+)
+
+// DownloadFromURL : Downloads the terraform binary and its hash from the source url
+func DownloadFromURL(installLocation string, mirrorURL string, tfversion string, versionPrefix string, goos string, goarch string) (string, error) {
+	pubKeyFilename := filepath.Join(installLocation, "/", PubKeyPrefix+PubKeyId+pubKeySuffix)
+	zipUrl := mirrorURL + tfversion + "/" + versionPrefix + tfversion + "_" + goos + "_" + goarch + ".zip"
+	hashUrl := mirrorURL + tfversion + "/" + versionPrefix + tfversion + "_SHA256SUMS"
+	hashSignatureUrl := mirrorURL + tfversion + "/" + versionPrefix + tfversion + "_SHA256SUMS." + PubKeyId + ".sig"
+
+	err := downloadPublicKey(installLocation, pubKeyFilename)
+	if err != nil {
+		log.Fatal("[Error]: Could not download public key file")
+	}
+
+	log.Println("Downloading ", zipUrl)
+	zipFilePath, err := downloadFromURL(installLocation, zipUrl)
+	if err != nil {
+		log.Fatal("[Error]: Could not download zip file")
+	}
+
+	log.Println("Downloading ", hashUrl)
+	hashFilePath, err := downloadFromURL(installLocation, hashUrl)
+	if err != nil {
+		log.Fatal("[Error]: Could not download hash file")
+	}
+
+	log.Println("Downloading ", hashSignatureUrl)
+	hashSigFilePath, err := downloadFromURL(installLocation, hashSignatureUrl)
+	if err != nil {
+		log.Fatal("[Error]: Could not download hash signature file")
+	}
+
+	publicKeyFile, err := os.Open(pubKeyFilename)
+	if err != nil {
+		log.Fatal("[Error]: Could not open the public key.", pubKeyFilename)
+	}
+
+	signatureFile, err := os.Open(hashSigFilePath)
+	if err != nil {
+		log.Fatal("[Error]: Could not open the signature file.", hashSigFilePath)
+	}
+
+	targetFile, err := os.Open(zipFilePath)
+	if err != nil {
+		log.Fatal("[Error]: Could not open the terraform binary for checksum verification.", zipFilePath)
+	}
+
+	hashFile, err := os.Open(hashFilePath)
+	if err != nil {
+		log.Fatal("[Error]: Could not open the terraform checksum file for signature verification.", hashFilePath)
+	}
+	verified := checkSignatureOfChecksums(publicKeyFile, hashFile, signatureFile)
+	if !verified {
+		return "", errors.New("signature of checksum files could not be verified")
```

Comment:
I'm not sure how much it matters - but I'm quite aware that if we download the public key, hash file and the zip file and the verification fails - all of these files are left in the install directory. Is it worth tidying them up after a failed download.

Reply:
In fact, the returned zip file path, once extracted by `Install()` will remove the zip file, but the checksum file will remain.

Reply:
I delete the files needed for the verification after the process now.